src/core/vector/qgsvectorlayer.cpp

```diff
@@ -189,6 +189,7 @@ QgsVectorLayer::QgsVectorLayer( const QString &vectorLayerPath,
       providerFlags |= QgsDataProvider::FlagLoadDefaultStyle;
     }
     setDataSourcePrivate( vectorLayerPath, baseName, providerKey, providerOptions, providerFlags );
+    emit dataSourceChanged();
```

Comment:
shouldn't this signal be emitted in setDataSourcePrivate ? (and removed from the other places where it's emitted)

Reply:
> The `emit` is done in the `setDataSource` method not in `setDataSourcePrivate` to do not have to duplicate the `emit` in every layer type implementation. 

Ok, then.

> Do you think I have to use `setDataSource` from `QgsMapLayer` instead of `setDataSourcePrivate` ?

Isn't that already the case? https://github.com/qgis/QGIS/blob/master/src/core/qgsmaplayer.cpp#L1719

Sorry, I'm a little confused: why do you need to call it in the constructor in the first place?

Reply:
In PR #42890, I have unvirtualize setDataSource, and I have replace setDataSoruce by setDataSourcePrivate. So unvirtualization has been firstly done in this PR and I probably forgot to apply this emit in #42890.

The setDataSource used in the QgsVectorLayer, before #42890, has `emit dataSourceChanged` before every `return`. I was thinking, it is important to keep `emit dataSourceChanged` here.

Reply:
Any reason why you are calling `setDataSourcePrivate()` instead of `setDataSource()` here? It would take care of emitting the signal directly.